Add extra debugging information to Bech32 decoding corruption tests.

jonathanknowles opened the pull request:
# Issue Number

Issue #324 

# Overview

- [x] I have added extra debugging information to the suite of Bech32 decoding corruption tests.

These tests:

1. corrupt a known-valid Bech32 string;
2. attempt to decode the corrupted string;
3. assert that decoding will fail.

Sample string:
```
       index of mutated char: 5
               original char: '1'
            replacement char: 'q'
             original string: "'14-b13zhcknjh"
            corrupted string: "'14-bq3zhcknjh"
``` 

Debugging information is only printed when QuickCheck finds a counterexample.

piotr-iohk commented on src/Cardano/Wallet/API/V1/ReifyWalletError.hs:
```diff
@@ -60,7 +61,12 @@ translateWalletLayerErrors ex = do
+importAddressError e = case e of
+    (ImportAddressError (Kernel.ImportAddressKeystoreNotFound _)) ->
+        V1.WalletNotFound
+    (ImportAddressAddressDecodingFailed _) ->
```
Does this mean that when address is invalid (e.g. too short, too long etc.) the `WalletNotFound` will be returned?

KtorZ replied:
Well, there's a bit of historical context here but indeed. The internal business logic is more expressive about errors than the outer API is capable of. And for backward-compatibility reasons, we haven't been able to extend the API. And that's why we have to do this mapping from Kernel errors to API errors. 

In the end, I map here to `WalletNotFound` for lack of a better one, and, for consistency with other errors mapping already in place. This happens indeed when whatever you submit as a wallet id is note a base58 encoded valid root id. Like I did in my example `/api/v1/wallets/123/addresses`. This returns a `WalletNotFound` (and `123` isn't a valid wallet id in a first place).

piotr-iohk replied:
ok, thanks. :+1:

piotr-iohk commented:
@KtorZ, this one needs to be ported to `cardano-sl`, no?

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that swaps adjacent characters.

Additionally, make it a precondition that the adjacent characters are actually
different to one another.

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that omits characters.

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that inserts characters.

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that mutates characters.

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that lower-cases characters.

Additionally, make it a precondition that the string has actually been
successfully corrupted.

jonathanknowles committed:
Add extra debug info to Bech32 corruption test that upper-cases characters.

Additionally, make it a precondition that the string has actually been
successfully corrupted.

rvl commented on lib/bech32/test/Codec/Binary/Bech32Spec.hs:
```diff
@@ -124,104 +126,132 @@ spec = do
+                        , " original string: " <> show originalString
+                        , "corrupted string: " <> show corruptedString ]
+                return $ counterexample description $
+                    char1 /= char2 ==>
```
Could this precondition be removed? Then change then line below to `shouldSatisfy (if char1 == char2 then isRight else isLeft)`

jonathanknowles replied:
@rvl wrote:

> Could this precondition be removed? Then change then line below to `shouldSatisfy (if char1 == char2 then isRight else isLeft)`

Actually, we do want this precondition, as we're **only** interested in the cases where corruption actually does happen. By making corruption a precondition, we guarantee to produce cases that we're interested in (or else QuickCheck will give up), where decoding is presumed to fail.

As it turns out, only a small proportion of tests (those that fail to produce a corrupted string) are discarded here.

Though perhaps you have another reason in mind for removing the precondition?

rvl commented on lib/bech32/test/Codec/Binary/Bech32Spec.hs:
```diff
@@ -124,104 +126,132 @@ spec = do
+                        , "      original string: " <> show originalString
+                        , "     corrupted string: " <> show corruptedString ]
+                return $ counterexample description $
+                    corruptedString /= originalString ==>
```
This precondition could be moved to the `shouldBe` easily enough.

jonathanknowles replied:
@rvl wrote:

> This precondition could be moved to the `shouldBe` easily enough.

I think this probably comes under the same discussion as https://github.com/input-output-hk/cardano-wallet/pull/327#discussion_r287939439

rvl reviewed (approved):
Looks good, I tried changing some of the expectations and running the properties to see what it looks like.
